### PR TITLE
fix(hooks): dedup concurrency-cancelled CI check-runs in the merge gate

### PR DESCRIPTION
**Pull request by WingedGuardian**
## Problem

`_pr_ci_status` (the CI gate on `gh pr merge`) classified every `statusCheckRollup`
entry independently, and `CANCELLED` is treated as red. When a
`concurrency: cancel-in-progress` run is superseded, its cancelled check-runs stay
attached to the head commit, so the gate read **red forever** even though the same
check's re-run concluded `SUCCESS` (measured on a real merge; cleared only by
manually re-running the cancelled job).

## Fix

Drop a `CANCELLED` CheckRun iff a check of the same `(name, workflowName)` identity
concluded `SUCCESS` on this head; otherwise it stays red. Pure set-membership —
**no time-ordering** (that surface caused a prior review loop).

Scoping by `workflowName` (not bare `name`) is deliberate and security-relevant:
this gate forces `gh pr merge --admin`, which bypasses GitHub's server-side
required-status-checks, so `_pr_ci_status` is the **sole** CI enforcement for every
merge it allows. A bare-name match would let a same-named `SUCCESS` from a
*different* workflow mask a genuinely-cancelled required check (wrong-green).
`(name, workflowName)` scopes the drop to a true same-job re-run — the only thing
`cancel-in-progress` produces — and fails closed when `workflowName` is absent.

## Safety

Wrong-green-impossible for realistic inputs: only `CANCELLED` (which carries no
verdict) is dropped; `FAILURE`/`TIMED_OUT` stay red; in-flight re-runs stay
pending; a cancel with no resolvable identity stays red. Residual (LOW,
documented-accept): a duplicate job `name:` inside the *same* gating workflow could
still collide — but that requires editing the CI definition itself, a strictly
higher bar, and such an actor can defeat CI more trivially. `gh pr view` exposes no
stronger identity field; a follow-up is filed for a `gh api …/check-runs` identity
rework.

## Testing

- TDD verify-RED observed (the pre-fix file fails the new tests for the right
  reason; independently confirmed by an adversarial security review that swapped
  the pre-fix file against the new tests).
- 10 new `TestPrCiStatusCancelSibling` cases incl. cross-workflow, no-workflowName,
  and fail-closed locks.
- `tests/test_hooks/` full: 1539 passed. `test_merge_review_gate.py`: 210 passed.
  `ruff` clean.
- Reviewed by genesis-architect + genesis-security-reviewer (two rounds); a HIGH
  (name-collision wrong-green) was found and fixed, no CRITICAL/HIGH remaining.

Follow-up: 121acade351b44f88f52130ed3785fff

### CODE REVIEW

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
## First-pass structural review

**Scope**: `scripts/hooks/git_push_guard.py` (+77/-2), `tests/test_hooks/test_merge_review_gate.py` (+110). No new files.

### Pattern violations (ERROR)
None found. No `nohup`, no raw SQL, no bare `except`, no mutable default args, no `/tmp/` usage, no `git add -A`/`--force`, no direct `task_states` writes — none of these patterns appear in the diff.

### Import hygiene (WARNING)
None. No new imports added; the new helpers (`_check_name`, `_ci_identity`) use only existing stdlib types (`tuple[str, str] | None`).

### Coverage (WARNING)
None. This is a change to an existing file, not a new module, so the "new file needs a test file" rule doesn't apply. The two new functions (`_check_name`, `_ci_identity`) and the new `_pr_ci_status` branch are exercised by the new `TestPrCiStatusCancelSibling` class (10 cases), which explicitly covers the identity-scoping edge cases (cross-workflow same-name, missing `workflowName`, in-flight re-run, unrelated pending/failure not swallowed, nameless entries). Verified no pre-existing test elsewhere in the file asserts the old "CANCELLED is always red" behavior in a way this change would silently invalidate — grepped for other `CANCELLED` references and found none outside the new test class.

### Documentation (NOTE)
None. Both new functions (`_check_name`, `_ci_identity`) have docstrings explaining purpose and rationale; `_pr_ci_status`'s docstring was updated to describe the new "red" classification for cancelled-without-sibling.

### Genesis conventions (WARNING)
- Commit message uses `fix(hooks):` — conventional prefix present.
- This is a hook script (`scripts/hooks/`) — diff adds no heavy third-party imports, consistent with the stdlib-only hook constraint.
- No `IntervalTrigger`/`$CLAUDE_PROJECT_DIR` usage in this diff.

### Minor observation (not filed as a finding — outside this checklist's categories)
`_ci_identity(c)` is computed twice for any CheckRun that is both in the SUCCESS pass (building `success_ids`) and later re-examined as a CANCELLED entry with a matching sibling — negligible cost given check-run counts are small, so not flagging as a real issue.

**Structural review: CLEAN — no pattern violations found.**



---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by WingedGuardian** [comment]
**Automated first-pass review**

Structural review: CLEAN — no pattern violations found.
- No nohup/direct-DB-insert/`/tmp`/`git add .`/force-push/bare-except/mutable-default patterns.
- Import hygiene clean: no new imports added; existing `git_push_guard.py` imports remain stdlib-only (matches the hook's declared "Stdlib-only" contract in its module docstring).
- Coverage: strong — 10 new cases in `TestPrCiStatusCancelSibling` cover the sibling-match happy path, cross-workflow non-match (security lock), missing-`workflowName` fail-closed, in-flight re-run, unrelated-failure non-laundering, `TIMED_OUT` scope lock, StatusContext/CheckRun shape collision, unrelated-pending preservation, and nameless-entry fail-safe.
- Documentation: `_check_name` and `_ci_identity` both have clear docstrings explaining purpose and rationale.
- Verified the safety claim in the PR body against the file directly: `CANCELLED` is indeed a member of `_CI_RED_CONCLUSIONS` (line 601), and the new sibling-drop branch runs *before* that red-classification check with its own `continue`, so the drop logic is additive/narrowing rather than replacing the default-red behavior — consistent with "fails closed unless a same-identity SUCCESS is found."
- Conventional commit prefix present (`fix(hooks): ...`).

PII/Secrets/Wording scan: CLEAN
- No API keys/tokens/credentials, no private IPs (10.x/192.168.x/100.64-127.x/fd00::/8), no SSH keys, no connection strings in the diff.
- No personal emails, home-directory usernames, revealing timezones, cloud account IDs, hardware identifiers, or private hostnames in the added lines.
- PR title/body/tests reference a prior PR number (#1420) and a short commit-sha-style incident label (`ec925917`) as internal context for the fix rationale — generic, non-exploitable, no private infrastructure named, no security-incident narrative beyond describing this repo's own CI-gate bug.



---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by WingedGuardian** [comment]
@codex review

File: scripts/hooks/git_push_guard.py [hunk 1]
```diff
@@ -654,19 +702,46 @@ def _pr_ci_status(pr_num: str, repo: str | None = None) -> tuple[str, list[str]]
...
+        name = _check_name(c)
         conclusion = c.get("conclusion")  # CheckRun
         status = c.get("status")  # CheckRun: QUEUED/IN_PROGRESS/COMPLETED/PENDING/…
         state = c.get("state")  # StatusContext: SUCCESS/FAILURE/PENDING/ERROR
         if conclusion in _CI_SKIP_CONCLUSIONS:
             saw_recognized = True
             continue
+        if conclusion in _CI_CANCEL_CONCLUSIONS:
+            ident = _ci_identity(c)
+            if ident is not None and ident in success_ids:
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/hooks/git_push_guard.py]: Addressed in b640d8b3: the drop is no longer bare set-membership — it now requires a same-identity (name, workflowName) SUCCESS whose completedAt is at-or-after the cancel (success_latest dict). SUCCESS-then-CANCELLED on an unchanged head therefore stays red (locked by test_success_then_cancel_on_unchanged_head_stays_red). The stricter tie-break you raise separately (P2, >= vs >) is under review.

File: scripts/hooks/git_push_guard.py [hunk 2]
```diff
@@ -654,33 +718,76 @@ def _pr_ci_status(pr_num: str, repo: str | None = None) -> tuple[str, list[str]]
...
         conclusion = c.get("conclusion")  # CheckRun
         status = c.get("status")  # CheckRun: QUEUED/IN_PROGRESS/COMPLETED/PENDING/…
         state = c.get("state")  # StatusContext: SUCCESS/FAILURE/PENDING/ERROR
         if conclusion in _CI_SKIP_CONCLUSIONS:
             saw_recognized = True
             continue
+        if conclusion in _CI_CANCEL_CONCLUSIONS:
+            ident = _ci_identity(c)
+            cts = (c.get("completedAt") or "").strip()
+            if ident is not None and cts and success_latest.get(ident, "") >= cts:
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply scripts/hooks/git_push_guard.py:758]: Consciously accepted (won't change). The drop requires success_latest[ident] >= cts where BOTH are terminal CheckRun completedAt values in whole-second UTC. A true tie (a superseded cancel and its superseding success completing in the same whole second) is not reachable in the concurrency-cancel scenario this targets: a cancel-in-progress run's completedAt is stamped when it is aborted (as the newer run starts), and the superseding SUCCESS completes strictly later — always a different second. On a hypothetical tie the run also genuinely passed on this head, so >= is benign, not a wrong-green in practice. This is a non-blocking P2 on the temporal surface past a 3-round escalation cap; per loop-termination discipline we're not restarting the full CI+review cycle for it. Noted in the tabled follow-up for the eventual gh-api check-runs identity rework.
